versioning/worker-versions.md

```diff
@@ -0,0 +1,174 @@
+which can automatically manage worker deployments.
+
+## Safe-rollout
+
+Intentionally avoiding deep into details here, the foundation laid here should enable
+us, as well as users, to more easily implement automated and safe rollout
+of changes to workers. A process like this one can become the de-facto experience:
+
+* If deploying a known-breaking change, new workflows are only processed by
+  the new version, existing open workflows are processed by old workers.
+* If deploying an expected-to-be-compatible change, new workflow _tasks_ are
+  attempted by new workers, and if they succeed, that workflow will subsequently
+  only be processed by new workers. If they fail, execution goes back to old
+  workers.
+
+# Using opaque whole-worker-identifiers as worker versions
+
+Thanks to @jlegrone for his great feedback encouraging us to revive this approach.
+
+We already have a field called `binaryChecksum` (or similar) in `WorkerOptions`
+today in SDKs. This field's intent is to uniquely identify a specific worker
+build. Since we believe versioning entire workers is the easiest solution, we
+can lean on this existing field.
+
+Deprecating the name `binaryChecksum`and replacing it with something more 
+accurate like `workerBuildId` would be appropriate, since most of our SDKs
+don't even produce a binary.
+
+With this approach, users who wish to make use of the improved versioning 
+features should set the `workerBuildId` during their CI process to a value that
+is unique for that worker's implementation. _Implementation_ rather than 
+_build_ is an important distinction here, since rebuilds of identical
+code should (ideally) produce workers wit the same ID. It is not necessary to
+do this, and it is varyingly difficult depending on the language, but when
+done it enables users to produce workers with the same ID from their CI builds
+if they didn't touch worker code, but something else (ex: The CI definition itself).
+This involves some work on the part of the user, but is more reliable than expecting
+worker authors to increment a version number.
+
+Unfortunately it is infeasible to automatically set an appropriate ID for many languages.
+
+## How is versioning accomplished with the `workerBuildId`?
+
+Every new worker ID will be assumed to be incompatible with the existing id(s)
+unless explicitly specified otherwise. New workers will not process *anything* 
+until explicitly enabled if they opt into build-id based versioning. Users who 
+wish to always have new workers begin processing new workflows can make an API 
+call immediately establishing the new worker as latest as part of their 
+deployment process (see below).
+
+In order to either (safely or not) migrate *open* workflows to a new version,
+users will explicitly indicate that an ID is compatible with some already-extant
+ID.
+
+Because this behavior is a breaking change, SDKs will need to add a
+`enableBuildIdBasedVersioning` boolean flag to their `WorkerOptions`.
```

Comment:
Could this be implied by the presence of a non empty workerBuildId?

Reply:
Yeah, it could be. I'm worried though it's maybe a bit non-obvious. Plus, if we deprecate the old checksum, you might want to set this but not actually opt in to the versioning behavior.